// ngx-mat-timepicker/esm5/lib/components/ngx-mat-timepicker-period/ngx-mat-timepicker-period.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-period/ngx-mat-timepicker-period.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { animate, sequence, style, transition, trigger } from "@angular/animations";
//
import { NgxMatTimepickerPeriods } from "../../models/ngx-mat-timepicker-periods.enum";
import { NgxMatTimepickerUnits } from "../../models/ngx-mat-timepicker-units.enum";
import { NgxMatTimepickerUtils } from "../../utils/ngx-mat-timepicker.utils";
//
import { DateTime } from "ts-luxon";
var NgxMatTimepickerPeriodComponent = /** @class */ (function () {
    function NgxMatTimepickerPeriodComponent() {
        this.isPeriodAvailable = true;
        this.periodChanged = new EventEmitter();
        this.timePeriod = NgxMatTimepickerPeriods;
    }
    /**
     * @return {?}
     */
    NgxMatTimepickerPeriodComponent.prototype.animationDone = /**
     * @return {?}
     */
    function () {
        this.isPeriodAvailable = true;
    };
    /**
     * @param {?} period
     * @return {?}
     */
    NgxMatTimepickerPeriodComponent.prototype.changePeriod = /**
     * @param {?} period
     * @return {?}
     */
    function (period) {
        this.isPeriodAvailable = this._isSwitchPeriodAvailable(period);
        if (this.isPeriodAvailable) {
            this.periodChanged.next(period);
        }
    };
    /**
     * @private
     * @param {?} period
     * @return {?}
     */
    NgxMatTimepickerPeriodComponent.prototype._getDisabledTimeByPeriod = /**
     * @private
     * @param {?} period
     * @return {?}
     */
    function (period) {
        switch (this.activeTimeUnit) {
            case NgxMatTimepickerUnits.HOUR:
                return NgxMatTimepickerUtils.disableHours(this.hours, {
                    min: this.minTime,
                    max: this.maxTime,
                    format: this.format,
                    period: period
                });
            case NgxMatTimepickerUnits.MINUTE:
                return NgxMatTimepickerUtils.disableMinutes(this.minutes, +this.selectedHour, {
                    min: this.minTime,
                    max: this.maxTime,
                    format: this.format,
                    period: period
                });
            default:
                throw new Error("no such NgxMatTimepickerUnits");
        }
    };
    /**
     * @private
     * @param {?} period
     * @return {?}
     */
    NgxMatTimepickerPeriodComponent.prototype._isSwitchPeriodAvailable = /**
     * @private
     * @param {?} period
     * @return {?}
     */
    function (period) {
        /** @type {?} */
        var time = this._getDisabledTimeByPeriod(period);
        return !time.every((/**
         * @param {?} t
         * @return {?}
         */
        function (t) { return t.disabled; }));
    };
    NgxMatTimepickerPeriodComponent.decorators = [
        { type: Component, args: [{
                    selector: "ngx-mat-timepicker-period",
                    template: "<div class=\"timepicker-period\">\n    <button class=\"timepicker-dial__item timepicker-period__btn\"\n            [ngClass]=\"{'active': selectedPeriod === timePeriod.AM}\"\n            (click)=\"changePeriod(timePeriod.AM)\"\n            type=\"button\">{{meridiems[0]}}</button>\n    <button class=\"timepicker-dial__item timepicker-period__btn\"\n            [ngClass]=\"{'active': selectedPeriod === timePeriod.PM}\"\n            (click)=\"changePeriod(timePeriod.PM)\"\n            type=\"button\">{{meridiems[1]}}</button>\n    <div class=\"timepicker-period__warning\"\n         [@scaleInOut]\n         (@scaleInOut.done)=\"animationDone()\"\n         *ngIf=\"!isPeriodAvailable\">\n        <p>Current time would be invalid in this period.</p>\n    </div>\n</div>\n",
                    animations: [
                        trigger("scaleInOut", [
                            transition(":enter", [
                                style({ transform: "scale(0)" }),
                                animate(".2s", style({ transform: "scale(1)" })),
                                sequence([
                                    animate("3s", style({ opacity: 1 })),
                                    animate(".3s", style({ opacity: 0 }))
                                ])
                            ])
                        ])
                    ],
                    styles: [".timepicker-period{display:flex;flex-direction:column;position:relative}.timepicker-period__btn{opacity:.5;padding:1px 3px;border:0;background-color:transparent;font-size:18px;font-weight:500;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;outline:0;border-radius:3px;transition:background-color .5s;color:inherit}.timepicker-period__btn.active{opacity:1}.timepicker-period__btn:focus{background-color:rgba(0,0,0,.07)}.timepicker-period__warning{padding:5px 10px;border-radius:3px;background-color:rgba(0,0,0,.55);position:absolute;width:200px;left:-20px;top:40px}.timepicker-period__warning>p{margin:0;font-size:12px}"]
                }] }
    ];
    NgxMatTimepickerPeriodComponent.propDecorators = {
        activeTimeUnit: [{ type: Input }],
        format: [{ type: Input }],
        hours: [{ type: Input }],
        maxTime: [{ type: Input }],
        meridiems: [{ type: Input }],
        minTime: [{ type: Input }],
        minutes: [{ type: Input }],
        periodChanged: [{ type: Output }],
        selectedHour: [{ type: Input }],
        selectedPeriod: [{ type: Input }]
    };
    return NgxMatTimepickerPeriodComponent;
}());
export { NgxMatTimepickerPeriodComponent };
if (false) {
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.activeTimeUnit;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.format;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.hours;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.isPeriodAvailable;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.maxTime;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.meridiems;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.minTime;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.minutes;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.periodChanged;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.selectedHour;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.selectedPeriod;
    /** @type {?} */
    NgxMatTimepickerPeriodComponent.prototype.timePeriod;
}
