// ngx-mat-timepicker/esm5/lib/components/ngx-mat-timepicker-toggle/ngx-mat-timepicker-toggle.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-toggle/ngx-mat-timepicker-toggle.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, Input } from "@angular/core";
import { NgxMatTimepickerToggleIconDirective } from "../../directives/ngx-mat-timepicker-toggle-icon.directive";
import { NgxMatTimepickerComponent } from "../ngx-mat-timepicker/ngx-mat-timepicker.component";
var NgxMatTimepickerToggleComponent = /** @class */ (function () {
    function NgxMatTimepickerToggleComponent() {
    }
    Object.defineProperty(NgxMatTimepickerToggleComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled === undefined ? this.timepicker.disabled : this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = value;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    NgxMatTimepickerToggleComponent.prototype.open = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.timepicker) {
            this.timepicker.open();
            event.stopPropagation();
        }
    };
    NgxMatTimepickerToggleComponent.decorators = [
        { type: Component, args: [{
                    selector: "ngx-mat-timepicker-toggle",
                    template: "<button class=\"ngx-mat-timepicker-toggle mat-elevation-z0\"\n        color=\"\"\n        mat-mini-fab\n        (click)=\"open($event)\"\n        [disabled]=\"disabled\"\n        type=\"button\">\n    <svg xmlns=\"http://www.w3.org/2000/svg\"\n         viewBox=\"0 0 24 24\"\n         width=\"24px\"\n         height=\"24px\"\n         *ngIf=\"!customIcon\">\n        <path d=\"M 12 2 C 6.4889971 2 2 6.4889971 2 12 C 2 17.511003                   6.4889971 22 12 22 C 17.511003 22 22 17.511003 22 12 C 22 6.4889971 17.511003 2 12 2 z M 12 4 C 16.430123 4 20 7.5698774 20 12 C 20 16.430123 16.430123 20 12 20 C 7.5698774 20 4 16.430123 4 12 C 4 7.5698774 7.5698774 4 12 4 z M 11 6 L 11 12.414062 L 15.292969 16.707031 L 16.707031 15.292969 L 13 11.585938 L 13 6 L 11 6 z\" />\n    </svg>\n\n    <ng-content select=\"[ngxMatTimepickerToggleIcon]\"></ng-content>\n</button>\n",
                    styles: ["button.ngx-mat-timepicker-toggle{background-color:transparent;text-align:center;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer;box-shadow:none}"]
                }] }
    ];
    NgxMatTimepickerToggleComponent.propDecorators = {
        disabled: [{ type: Input }],
        customIcon: [{ type: ContentChild, args: [NgxMatTimepickerToggleIconDirective, { static: true },] }],
        timepicker: [{ type: Input, args: ["for",] }]
    };
    return NgxMatTimepickerToggleComponent;
}());
export { NgxMatTimepickerToggleComponent };
if (false) {
    /** @type {?} */
    NgxMatTimepickerToggleComponent.prototype.customIcon;
    /** @type {?} */
    NgxMatTimepickerToggleComponent.prototype.timepicker;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerToggleComponent.prototype._disabled;
}
