// ngx-mat-timepicker/esm5/lib/components/ngx-mat-timepicker-hours-face/ngx-mat-timepicker-hours-face.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-hours-face/ngx-mat-timepicker-hours-face.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from "@angular/core";
import { NgxMatTimepickerUtils } from "../../utils/ngx-mat-timepicker.utils";
//
import { DateTime } from "ts-luxon";
var NgxMatTimepickerHoursFaceDirective = /** @class */ (function () {
    function NgxMatTimepickerHoursFaceDirective() {
        this.hourChange = new EventEmitter();
        this.hourSelected = new EventEmitter();
        this.hoursList = [];
        this._color = "primary";
        this._format = 24;
    }
    Object.defineProperty(NgxMatTimepickerHoursFaceDirective.prototype, "color", {
        get: /**
         * @return {?}
         */
        function () {
            return this._color;
        },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            this._color = newValue;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NgxMatTimepickerHoursFaceDirective.prototype, "format", {
        get: /**
         * @return {?}
         */
        function () {
            return this._format;
        },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            this._format = newValue;
            this.hoursList = NgxMatTimepickerUtils.getHours(this._format);
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @param {?} time
     * @return {?}
     */
    NgxMatTimepickerHoursFaceDirective.prototype.onTimeSelected = /**
     * @param {?} time
     * @return {?}
     */
    function (time) {
        this.hourSelected.next(time);
    };
    NgxMatTimepickerHoursFaceDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[ngxMatTimepickerHoursFace]"
                },] }
    ];
    /** @nocollapse */
    NgxMatTimepickerHoursFaceDirective.ctorParameters = function () { return []; };
    NgxMatTimepickerHoursFaceDirective.propDecorators = {
        color: [{ type: Input }],
        format: [{ type: Input }],
        hourChange: [{ type: Output }],
        hourSelected: [{ type: Output }],
        maxTime: [{ type: Input }],
        minTime: [{ type: Input }],
        selectedHour: [{ type: Input }]
    };
    return NgxMatTimepickerHoursFaceDirective;
}());
export { NgxMatTimepickerHoursFaceDirective };
if (false) {
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.hourChange;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.hourSelected;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.hoursList;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.maxTime;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.minTime;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.selectedHour;
    /**
     * @type {?}
     * @protected
     */
    NgxMatTimepickerHoursFaceDirective.prototype._color;
    /**
     * @type {?}
     * @protected
     */
    NgxMatTimepickerHoursFaceDirective.prototype._format;
}
