// smp-core-utils/bundles/smp-core-utils.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/material/core'), require('lodash'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('smp-core-utils', ['exports', '@angular/core', '@angular/material/core', 'lodash', 'rxjs'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['smp-core-utils'] = {}, global.ng.core, global.ng.material.core, global.lodash, global.rxjs));
}(this, (function (exports, core$1, core, lodash, rxjs) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var smpSep = "/";
    var smpDay = "d";
    var smpDayDoubled = "dd";
    var smpMonth = "M";
    var smpMonthDoubled = "MM";
    var smpYear = "yyyy";
    var smpCurrentYear = new Date().getFullYear();
    var smpCentury = Math.round(smpCurrentYear / 100);
    var SMP_LOCALE_IT_IT = "it-IT";
    var SMP_LOCALE_EN_GB = "en-GB";
    var SMP_LOCALE_EN_US = "en-US";
    var SMP_FORMATS = {
        "it-IT": "" + smpDayDoubled + smpSep + smpMonthDoubled + smpSep + smpYear,
        "en-GB": "" + smpDayDoubled + smpSep + smpMonthDoubled + smpSep + smpYear,
        "en-US": "" + smpMonth + smpSep + smpDay + smpSep + smpYear,
    };
    var SMP_DATE_FORMATS_IT = {
        parse: {
            dateInput: { month: "numeric", year: "numeric", day: "numeric" }
        },
        display: {
            dateInput: SMP_FORMATS[SMP_LOCALE_IT_IT],
            monthYearLabel: "" + smpMonthDoubled + smpSep + smpYear,
            dateA11yLabel: SMP_FORMATS[SMP_LOCALE_IT_IT],
            monthYearA11yLabel: "" + smpMonthDoubled + smpSep + smpYear
        }
    };
    var SMP_DATE_FORMATS_EN_GB = {
        parse: {
            dateInput: { month: "numeric", year: "numeric", day: "numeric" }
        },
        display: {
            dateInput: SMP_FORMATS[SMP_LOCALE_EN_GB],
            monthYearLabel: "" + smpMonthDoubled + smpSep + smpYear,
            dateA11yLabel: SMP_FORMATS[SMP_LOCALE_EN_GB],
            monthYearA11yLabel: "" + smpMonthDoubled + smpSep + smpYear
        }
    };
    var SMP_DATE_FORMATS_EN_US = {
        parse: {
            dateInput: { month: "numeric", year: "numeric", day: "numeric" }
        },
        display: {
            dateInput: SMP_FORMATS[SMP_LOCALE_EN_US],
            monthYearLabel: "" + smpMonth + smpSep + smpYear,
            dateA11yLabel: SMP_FORMATS[SMP_LOCALE_EN_US],
            monthYearA11yLabel: "" + smpMonth + smpSep + smpYear
        }
    };
    var SmpMatDateLocale = /** @class */ (function (_super) {
        __extends(SmpMatDateLocale, _super);
        function SmpMatDateLocale() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.display = SMP_DATE_FORMATS_EN_GB.display;
            return _this;
        }
        Object.defineProperty(SmpMatDateLocale.prototype, "currentLocale", {
            get: function () {
                return this.locale;
            },
            enumerable: false,
            configurable: true
        });
        SmpMatDateLocale.prototype.format = function (date, _displayFormat_) {
            var day = date.getDate();
            var month = date.getMonth() + 1;
            var year = date.getFullYear();
            switch (this.display.dateInput) {
                case SMP_DATE_FORMATS_IT.display.dateInput:
                    return this._addLeadingZeroes(day) + smpSep + this._addLeadingZeroes(month) + smpSep + year;
                case SMP_DATE_FORMATS_EN_GB.display.dateInput:
                    return this._addLeadingZeroes(day) + smpSep + this._addLeadingZeroes(month) + smpSep + year;
                case SMP_DATE_FORMATS_EN_US.display.dateInput:
                    return month + smpSep + day + smpSep + year;
                default:
                    var formattedDate = "";
                    try {
                        formattedDate = date.toLocaleDateString(SMP_LOCALE_EN_GB);
                    }
                    catch (e) {
                        console.warn("Catched error while printing ", e);
                        formattedDate = date.toDateString();
                    }
                    return formattedDate;
            }
        };
        SmpMatDateLocale.prototype.parse = function (value) {
            if (lodash.includes([SMP_DATE_FORMATS_IT.display.dateInput, SMP_DATE_FORMATS_EN_GB.display.dateInput], this.display.dateInput)) {
                if (typeof value === typeof "" && value.length > 0) {
                    var str = value.split(smpSep);
                    if (str.length < 3) {
                        return new Date("smp");
                    }
                    if (str[2].length < 4) {
                        str[2] = smpCentury + str[2];
                    }
                    if (str[2].length !== 4) {
                        return new Date("smp");
                    }
                    var year = +str[2];
                    var month = +str[1] - 1;
                    var day = +str[0];
                    return new Date(year, month, day);
                }
                return null;
            }
            else if (this.display.dateInput === SMP_DATE_FORMATS_EN_US.display.dateInput) {
                if (typeof value === typeof "" && value.length > 0) {
                    var str = value.split(smpSep);
                    if (str.length < 3) {
                        return new Date("smp");
                    }
                    if (str[2].length < 4) {
                        str[2] = smpCentury + str[2];
                    }
                    if (str[2].length !== 4) {
                        return new Date("smp");
                    }
                    var year = +str[2];
                    var month = +str[0] - 1;
                    var day = +str[1];
                    return new Date(year, month, day);
                }
                return null;
            }
            else {
                console.warn("Date input format not valid." +
                    " Be sure to use SMP_DATE_FORMATS" +
                    " or extend this class with your own 'parser' and format 'methods'");
                console.info("For a complete guide visit github.com/tonysamperi/ngx-mat-lib");
                return null;
            }
        };
        SmpMatDateLocale.prototype.setLocale = function (locale) {
            switch (locale) {
                case SMP_LOCALE_IT_IT:
                    this.display = SMP_DATE_FORMATS_IT.display;
                    break;
                case SMP_LOCALE_EN_GB:
                    this.display = SMP_DATE_FORMATS_EN_GB.display;
                    break;
                case SMP_LOCALE_EN_US:
                    this.display = SMP_DATE_FORMATS_EN_US.display;
                    break;
                default:
                    this.display = SMP_DATE_FORMATS_EN_GB.display;
                    break;
            }
            _super.prototype.setLocale.call(this, locale);
        };
        SmpMatDateLocale.prototype._addLeadingZeroes = function (n) {
            return ("0" + n).slice(-2);
        };
        return SmpMatDateLocale;
    }(core.NativeDateAdapter));
    SmpMatDateLocale.decorators = [
        { type: core$1.Injectable }
    ];

    var SmpDomService = /** @class */ (function () {
        function SmpDomService(_componentFactoryResolver, _appRef, _injector) {
            this._componentFactoryResolver = _componentFactoryResolver;
            this._appRef = _appRef;
            this._injector = _injector;
            this._refs = {};
        }
        SmpDomService.prototype.appendComp = function (component, key, target, injector, waiter) {
            if (target === void 0) { target = "body"; }
            if (injector === void 0) { injector = this._injector; }
            if (waiter === void 0) { waiter = rxjs.timer(); }
            var componentRef = this._componentFactoryResolver
                .resolveComponentFactory(component)
                .create(injector);
            this._appRef.attachView(componentRef.hostView);
            var domElem = componentRef.hostView
                .rootNodes[0];
            if (!!key) {
                !!this._refs[key] && console.warn("A ref with this key already exists and will be overwritten");
                this._refs[key] = componentRef;
            }
            else {
                console.warn("Component appended, but key wasn't provided. It won't be possible to remove it from the DOM automagically");
            }
            waiter.subscribe(function () {
                var targetEl = (typeof target === typeof ""
                    ? document.querySelector(target)
                    : target instanceof core$1.ElementRef ? target.nativeElement : target) || document.body;
                targetEl.appendChild(domElem);
            });
            return componentRef;
        };
        SmpDomService.prototype.appendComps = function (compsData, injector, waiter) {
            var _this = this;
            compsData.forEach(function (compData) { return _this.appendComp(compData.comp, compData.key, compData.target, injector, waiter); });
        };
        SmpDomService.prototype.createInjector = function (token, value, viewContainerRef) {
            var providers = [
                { provide: token, useValue: value }
            ];
            return core$1.Injector.create({
                parent: viewContainerRef ? viewContainerRef.injector : this._injector,
                providers: providers
            });
        };
        SmpDomService.prototype.removeComp = function (key) {
            if (!!this._refs[key]) {
                var componentRef = this._refs[key];
                this._appRef.detachView(componentRef.hostView);
                componentRef.destroy();
                return !0;
            }
            else {
                console.warn("No ref found with key " + key);
                return !1;
            }
        };
        return SmpDomService;
    }());
    SmpDomService.decorators = [
        { type: core$1.Injectable }
    ];
    SmpDomService.ctorParameters = function () { return [
        { type: core$1.ComponentFactoryResolver },
        { type: core$1.ApplicationRef },
        { type: core$1.Injector }
    ]; };

    var SmpCoreUtilsModule = /** @class */ (function () {
        function SmpCoreUtilsModule() {
        }
        SmpCoreUtilsModule.forRoot = function () {
            return {
                ngModule: SmpCoreUtilsModule,
                providers: [
                    SmpDomService
                ]
            };
        };
        return SmpCoreUtilsModule;
    }());
    SmpCoreUtilsModule.decorators = [
        { type: core$1.NgModule, args: [{},] }
    ];

    exports.SMP_DATE_FORMATS_EN_GB = SMP_DATE_FORMATS_EN_GB;
    exports.SMP_DATE_FORMATS_EN_US = SMP_DATE_FORMATS_EN_US;
    exports.SMP_DATE_FORMATS_IT = SMP_DATE_FORMATS_IT;
    exports.SMP_FORMATS = SMP_FORMATS;
    exports.SMP_LOCALE_EN_GB = SMP_LOCALE_EN_GB;
    exports.SMP_LOCALE_EN_US = SMP_LOCALE_EN_US;
    exports.SMP_LOCALE_IT_IT = SMP_LOCALE_IT_IT;
    exports.SmpCoreUtilsModule = SmpCoreUtilsModule;
    exports.SmpDomService = SmpDomService;
    exports.SmpMatDateLocale = SmpMatDateLocale;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
