// ngx-mat-timepicker/esm5/lib/directives/ngx-mat-timepicker.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ngx-mat-timepicker.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, HostBinding, Inject, Input, Optional } from "@angular/core";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
//
import { NgxMatTimepickerComponent } from "../components/ngx-mat-timepicker/ngx-mat-timepicker.component";
import { NgxMatTimepickerAdapter } from "../services/ngx-mat-timepicker-adapter";
import { NGX_MAT_TIMEPICKER_LOCALE } from "../tokens/ngx-mat-timepicker-time-locale.token";
import { Subject } from "rxjs";
import { takeUntil } from "rxjs/operators";
import { CdkOverlayOrigin } from "@angular/cdk/overlay";
import { MatFormField } from "@angular/material/form-field";
var NgxMatTimepickerDirective = /** @class */ (function () {
    function NgxMatTimepickerDirective(_elementRef, _matFormField, _locale) {
        this._elementRef = _elementRef;
        this._matFormField = _matFormField;
        this._locale = _locale;
        // TODO: IMPROVE DETECTING (INJECT) MAT-FORM-FIELD IF PRESENT
        this.cdkOverlayOrigin = new CdkOverlayOrigin(this._matFormField ? this._matFormField.getConnectedOverlayOrigin() : this._elementRef);
        this._format = 12;
        this._subsCtrl$ = new Subject();
        this._value = "";
        this.onTouched = (/**
         * @return {?}
         */
        function () {
        });
        this._onChange = (/**
         * @return {?}
         */
        function () {
        });
    }
    Object.defineProperty(NgxMatTimepickerDirective.prototype, "element", {
        get: /**
         * @return {?}
         */
        function () {
            return this._elementRef && this._elementRef.nativeElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NgxMatTimepickerDirective.prototype, "format", {
        get: /**
         * @return {?}
         */
        function () {
            return this._format;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._format = +value === 24 ? 24 : 12;
            /** @type {?} */
            var isDynamicallyChanged = value && (this._previousFormat && this._previousFormat !== this._format);
            if (isDynamicallyChanged) {
                this.value = this._value;
                this._timepicker.updateTime(this._value);
            }
            this._previousFormat = this._format;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NgxMatTimepickerDirective.prototype, "max", {
        get: /**
         * @return {?}
         */
        function () {
            return this._max;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (typeof value === "string") {
                this._max = NgxMatTimepickerAdapter.parseTime(value, { locale: this._locale, format: this.format });
                return;
            }
            this._max = value;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NgxMatTimepickerDirective.prototype, "min", {
        get: /**
         * @return {?}
         */
        function () {
            return this._min;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (typeof value === "string") {
                this._min = NgxMatTimepickerAdapter.parseTime(value, { locale: this._locale, format: this.format });
                return;
            }
            this._min = value;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NgxMatTimepickerDirective.prototype, "timepicker", {
        set: /**
         * @param {?} picker
         * @return {?}
         */
        function (picker) {
            this._registerTimepicker(picker);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NgxMatTimepickerDirective.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            if (!this._value) {
                return "";
            }
            return NgxMatTimepickerAdapter.toLocaleTimeString(this._value, { format: this.format, locale: this._locale });
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!value) {
                this._value = "";
                this._updateInputValue();
                return;
            }
            /** @type {?} */
            var time = NgxMatTimepickerAdapter.formatTime(value, { locale: this._locale, format: this.format });
            /** @type {?} */
            var isAvailable = NgxMatTimepickerAdapter.isTimeAvailable(time, (/** @type {?} */ (this._min)), (/** @type {?} */ (this._max)), "minutes", this._timepicker.minutesGap, this._format);
            if (isAvailable) {
                this._value = time;
                this._updateInputValue();
                return;
            }
            console.warn("Selected time doesn't match min or max value");
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NgxMatTimepickerDirective.prototype, "_defaultTime", {
        set: /**
         * @private
         * @param {?} time
         * @return {?}
         */
        function (time) {
            this._timepicker.defaultTime = NgxMatTimepickerAdapter.formatTime(time, {
                locale: this._locale,
                format: this.format
            });
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.value && changes.value.currentValue) {
            this._defaultTime = changes.value.currentValue;
        }
    };
    /**
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._unregisterTimepicker();
        this._subsCtrl$.next();
        this._subsCtrl$.complete();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.onClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (!this.disableClick) {
            this._timepicker.open();
            event.stopPropagation();
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this._onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.setDisabledState = /**
     * @param {?} isDisabled
     * @return {?}
     */
    function (isDisabled) {
        this.disabled = isDisabled;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.updateValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
        this._onChange(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.value = value;
        if (value) {
            this._defaultTime = value;
        }
    };
    /**
     * @private
     * @param {?} picker
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype._registerTimepicker = /**
     * @private
     * @param {?} picker
     * @return {?}
     */
    function (picker) {
        var _this = this;
        if (picker) {
            this._timepicker = picker;
            this._timepicker.registerInput(this);
            this._timepicker.timeSet
                .pipe(takeUntil(this._subsCtrl$))
                .subscribe((/**
             * @param {?} time
             * @return {?}
             */
            function (time) {
                _this.value = time;
                _this._onChange(_this.value);
                _this.onTouched();
                _this._defaultTime = _this._value;
            }));
        }
        else {
            throw new Error("NgxMatTimepickerComponent is not defined." +
                " Please make sure you passed the timepicker to ngxMatTimepicker directive");
        }
    };
    /**
     * @private
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype._unregisterTimepicker = /**
     * @private
     * @return {?}
     */
    function () {
        if (this._timepicker) {
            this._timepicker.unregisterInput();
        }
    };
    /**
     * @private
     * @return {?}
     */
    NgxMatTimepickerDirective.prototype._updateInputValue = /**
     * @private
     * @return {?}
     */
    function () {
        this._elementRef.nativeElement.value = this.value;
    };
    NgxMatTimepickerDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[ngxMatTimepicker]",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: NgxMatTimepickerDirective,
                            multi: true
                        }
                    ],
                    // tslint:disable-next-line:no-host-metadata-property
                    host: {
                        "[disabled]": "disabled",
                        "(blur)": "onTouched()",
                    },
                },] }
    ];
    /** @nocollapse */
    NgxMatTimepickerDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: MatFormField, decorators: [{ type: Optional }, { type: Inject, args: [MatFormField,] }] },
        { type: String, decorators: [{ type: Inject, args: [NGX_MAT_TIMEPICKER_LOCALE,] }] }
    ]; };
    NgxMatTimepickerDirective.propDecorators = {
        format: [{ type: Input }],
        max: [{ type: Input }],
        min: [{ type: Input }],
        timepicker: [{ type: Input, args: ["ngxMatTimepicker",] }],
        value: [{ type: Input }],
        cdkOverlayOrigin: [{ type: HostBinding, args: ["attr.cdkOverlayOrigin",] }],
        disableClick: [{ type: Input }],
        disabled: [{ type: Input }],
        onClick: [{ type: HostListener, args: ["click", ["$event"],] }],
        updateValue: [{ type: HostListener, args: ["change", ["$event"],] }]
    };
    return NgxMatTimepickerDirective;
}());
export { NgxMatTimepickerDirective };
if (false) {
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.cdkOverlayOrigin;
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.disableClick;
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.disabled;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._format;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._max;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._min;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._previousFormat;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._subsCtrl$;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._timepicker;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._value;
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._onChange;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._matFormField;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._locale;
}
