// ngx-mat-timepicker/esm5/lib/directives/ngx-mat-timepicker-autofocus.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ngx-mat-timepicker-autofocus.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Inject, Input, Optional } from "@angular/core";
import { DOCUMENT } from "@angular/common";
var NgxMatTimepickerAutofocusDirective = /** @class */ (function () {
    function NgxMatTimepickerAutofocusDirective(_element, document) {
        this._element = _element;
        this.document = document;
        this._activeElement = this.document.activeElement;
    }
    /**
     * @return {?}
     */
    NgxMatTimepickerAutofocusDirective.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.isFocusActive) {
            // To avoid ExpressionChangedAfterItHasBeenCheckedError;
            setTimeout((/**
             * @return {?}
             */
            function () { return _this._element.nativeElement.focus({ preventScroll: true }); }));
        }
    };
    /**
     * @return {?}
     */
    NgxMatTimepickerAutofocusDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // To avoid ExpressionChangedAfterItHasBeenCheckedError;
        setTimeout((/**
         * @return {?}
         */
        function () { return _this._activeElement.focus({ preventScroll: true }); }));
    };
    NgxMatTimepickerAutofocusDirective.decorators = [
        { type: Directive, args: [{
                    selector: "[ngxMatTimepickerAutofocus]"
                },] }
    ];
    /** @nocollapse */
    NgxMatTimepickerAutofocusDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [DOCUMENT,] }] }
    ]; };
    NgxMatTimepickerAutofocusDirective.propDecorators = {
        isFocusActive: [{ type: Input, args: ["ngxMatTimepickerAutofocus",] }]
    };
    return NgxMatTimepickerAutofocusDirective;
}());
export { NgxMatTimepickerAutofocusDirective };
if (false) {
    /** @type {?} */
    NgxMatTimepickerAutofocusDirective.prototype.isFocusActive;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerAutofocusDirective.prototype._activeElement;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerAutofocusDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerAutofocusDirective.prototype.document;
}
