// ngx-mat-timepicker/esm5/lib/components/ngx-mat-timepicker-12-hours-face/ngx-mat-timepicker-12-hours-face.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-12-hours-face/ngx-mat-timepicker-12-hours-face.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __extends } from "tslib";
import { ChangeDetectionStrategy, Component, Input } from "@angular/core";
//
import { NgxMatTimepickerHoursFaceDirective } from "../ngx-mat-timepicker-hours-face/ngx-mat-timepicker-hours-face.directive";
import { NgxMatTimepickerPeriods } from "../../models/ngx-mat-timepicker-periods.enum";
import { NgxMatTimepickerUtils } from "../../utils/ngx-mat-timepicker.utils";
var NgxMatTimepicker12HoursFaceComponent = /** @class */ (function (_super) {
    __extends(NgxMatTimepicker12HoursFaceComponent, _super);
    function NgxMatTimepicker12HoursFaceComponent() {
        var _this = _super.call(this) || this;
        _this.format = 12;
        return _this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxMatTimepicker12HoursFaceComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.period && changes.period.currentValue) {
            this.hoursList = NgxMatTimepickerUtils.disableHours(this.hoursList, {
                min: this.minTime,
                max: this.maxTime,
                format: this.format,
                period: this.period
            });
        }
    };
    NgxMatTimepicker12HoursFaceComponent.decorators = [
        { type: Component, args: [{
                    selector: "ngx-mat-timepicker-12-hours-face",
                    template: "<ngx-mat-timepicker-face [color]=\"color\"\n                     [selectedTime]=\"selectedHour\"\n                     [faceTime]=\"hoursList\"\n                     (timeChange)=\"hourChange.next($event)\"\n                     (timeSelected)=\"onTimeSelected($event)\"></ngx-mat-timepicker-face>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    NgxMatTimepicker12HoursFaceComponent.ctorParameters = function () { return []; };
    NgxMatTimepicker12HoursFaceComponent.propDecorators = {
        period: [{ type: Input }]
    };
    return NgxMatTimepicker12HoursFaceComponent;
}(NgxMatTimepickerHoursFaceDirective));
export { NgxMatTimepicker12HoursFaceComponent };
if (false) {
    /** @type {?} */
    NgxMatTimepicker12HoursFaceComponent.prototype.period;
}
