// ngx-mat-timepicker/esm5/lib/components/ngx-mat-timepicker-24-hours-face/ngx-mat-timepicker-24-hours-face.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-24-hours-face/ngx-mat-timepicker-24-hours-face.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __extends } from "tslib";
import { ChangeDetectionStrategy, Component } from "@angular/core";
//
import { NgxMatTimepickerHoursFaceDirective } from "../ngx-mat-timepicker-hours-face/ngx-mat-timepicker-hours-face.directive";
import { NgxMatTimepickerUtils } from "../../utils/ngx-mat-timepicker.utils";
var NgxMatTimepicker24HoursFaceComponent = /** @class */ (function (_super) {
    __extends(NgxMatTimepicker24HoursFaceComponent, _super);
    function NgxMatTimepicker24HoursFaceComponent() {
        var _this = _super.call(this) || this;
        _this.format = 24;
        return _this;
    }
    /**
     * @return {?}
     */
    NgxMatTimepicker24HoursFaceComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.hoursList = NgxMatTimepickerUtils.disableHours(this.hoursList, {
            min: this.minTime,
            max: this.maxTime,
            format: this.format
        });
    };
    NgxMatTimepicker24HoursFaceComponent.decorators = [
        { type: Component, args: [{
                    selector: "ngx-mat-timepicker-24-hours-face",
                    template: "<ngx-mat-timepicker-face [color]=\"color\"\n                     [selectedTime]=\"selectedHour\"\n                     [faceTime]=\"hoursList\"\n                     [format]=\"format\"\n                     (timeChange)=\"hourChange.next($event)\"\n                     (timeSelected)=\"onTimeSelected($event)\"></ngx-mat-timepicker-face>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    NgxMatTimepicker24HoursFaceComponent.ctorParameters = function () { return []; };
    return NgxMatTimepicker24HoursFaceComponent;
}(NgxMatTimepickerHoursFaceDirective));
export { NgxMatTimepicker24HoursFaceComponent };
