// ngx-mat-timepicker/esm2015/lib/components/ngx-mat-timepicker-hours-face/ngx-mat-timepicker-hours-face.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-hours-face/ngx-mat-timepicker-hours-face.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, EventEmitter, Input, Output } from "@angular/core";
import { NgxMatTimepickerUtils } from "../../utils/ngx-mat-timepicker.utils";
//
import { DateTime } from "ts-luxon";
export class NgxMatTimepickerHoursFaceDirective {
    constructor() {
        this.hourChange = new EventEmitter();
        this.hourSelected = new EventEmitter();
        this.hoursList = [];
        this._color = "primary";
        this._format = 24;
    }
    /**
     * @param {?} newValue
     * @return {?}
     */
    set color(newValue) {
        this._color = newValue;
    }
    /**
     * @return {?}
     */
    get color() {
        return this._color;
    }
    /**
     * @param {?} newValue
     * @return {?}
     */
    set format(newValue) {
        this._format = newValue;
        this.hoursList = NgxMatTimepickerUtils.getHours(this._format);
    }
    /**
     * @return {?}
     */
    get format() {
        return this._format;
    }
    /**
     * @param {?} time
     * @return {?}
     */
    onTimeSelected(time) {
        this.hourSelected.next(time);
    }
}
NgxMatTimepickerHoursFaceDirective.decorators = [
    { type: Directive, args: [{
                selector: "[ngxMatTimepickerHoursFace]"
            },] }
];
/** @nocollapse */
NgxMatTimepickerHoursFaceDirective.ctorParameters = () => [];
NgxMatTimepickerHoursFaceDirective.propDecorators = {
    color: [{ type: Input }],
    format: [{ type: Input }],
    hourChange: [{ type: Output }],
    hourSelected: [{ type: Output }],
    maxTime: [{ type: Input }],
    minTime: [{ type: Input }],
    selectedHour: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.hourChange;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.hourSelected;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.hoursList;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.maxTime;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.minTime;
    /** @type {?} */
    NgxMatTimepickerHoursFaceDirective.prototype.selectedHour;
    /**
     * @type {?}
     * @protected
     */
    NgxMatTimepickerHoursFaceDirective.prototype._color;
    /**
     * @type {?}
     * @protected
     */
    NgxMatTimepickerHoursFaceDirective.prototype._format;
}
