// ngx-mat-timepicker/esm2015/lib/components/ngx-mat-timepicker-minutes-face/ngx-mat-timepicker-minutes-face.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-minutes-face/ngx-mat-timepicker-minutes-face.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { NgxMatTimepickerUnits } from "../../models/ngx-mat-timepicker-units.enum";
import { NgxMatTimepickerPeriods } from "../../models/ngx-mat-timepicker-periods.enum";
import { NgxMatTimepickerUtils } from "../../utils/ngx-mat-timepicker.utils";
//
import { DateTime } from "ts-luxon";
export class NgxMatTimepickerMinutesFaceComponent {
    constructor() {
        this.minuteChange = new EventEmitter();
        this.minutesList = [];
        this.timeUnit = NgxMatTimepickerUnits;
        this._color = "primary";
    }
    /**
     * @param {?} newValue
     * @return {?}
     */
    set color(newValue) {
        this._color = newValue;
    }
    /**
     * @return {?}
     */
    get color() {
        return this._color;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.period && changes.period.currentValue) {
            /** @type {?} */
            const minutes = NgxMatTimepickerUtils.getMinutes(this.minutesGap);
            this.minutesList = NgxMatTimepickerUtils.disableMinutes(minutes, this.selectedHour, {
                min: this.minTime,
                max: this.maxTime,
                format: this.format,
                period: this.period
            });
        }
    }
}
NgxMatTimepickerMinutesFaceComponent.decorators = [
    { type: Component, args: [{
                selector: "ngx-mat-timepicker-minutes-face",
                template: "<ngx-mat-timepicker-face [color]=\"color\"\n                     [faceTime]=\"minutesList\"\n                     [selectedTime]=\"selectedMinute\"\n                     [minutesGap]=\"minutesGap\"\n                     (timeChange)=\"minuteChange.next($event)\"\n                     [unit]=\"timeUnit.MINUTE\"></ngx-mat-timepicker-face>\n"
            }] }
];
/** @nocollapse */
NgxMatTimepickerMinutesFaceComponent.ctorParameters = () => [];
NgxMatTimepickerMinutesFaceComponent.propDecorators = {
    color: [{ type: Input }],
    format: [{ type: Input }],
    maxTime: [{ type: Input }],
    minTime: [{ type: Input }],
    minuteChange: [{ type: Output }],
    minutesGap: [{ type: Input }],
    period: [{ type: Input }],
    selectedHour: [{ type: Input }],
    selectedMinute: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.format;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.maxTime;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minTime;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minuteChange;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minutesGap;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minutesList;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.period;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.selectedHour;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.selectedMinute;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.timeUnit;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerMinutesFaceComponent.prototype._color;
}
