// ngx-mat-timepicker/esm2015/lib/directives/ngx-mat-timepicker.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ngx-mat-timepicker.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, HostBinding, Inject, Input, Optional } from "@angular/core";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
//
import { NgxMatTimepickerComponent } from "../components/ngx-mat-timepicker/ngx-mat-timepicker.component";
import { NgxMatTimepickerAdapter } from "../services/ngx-mat-timepicker-adapter";
import { NGX_MAT_TIMEPICKER_LOCALE } from "../tokens/ngx-mat-timepicker-time-locale.token";
import { Subject } from "rxjs";
import { takeUntil } from "rxjs/operators";
import { CdkOverlayOrigin } from "@angular/cdk/overlay";
import { MatFormField } from "@angular/material/form-field";
export class NgxMatTimepickerDirective {
    /**
     * @param {?} _elementRef
     * @param {?} _matFormField
     * @param {?} _locale
     */
    constructor(_elementRef, _matFormField, _locale) {
        this._elementRef = _elementRef;
        this._matFormField = _matFormField;
        this._locale = _locale;
        // TODO: IMPROVE DETECTING (INJECT) MAT-FORM-FIELD IF PRESENT
        this.cdkOverlayOrigin = new CdkOverlayOrigin(this._matFormField ? this._matFormField.getConnectedOverlayOrigin() : this._elementRef);
        this._format = 12;
        this._subsCtrl$ = new Subject();
        this._value = "";
        this.onTouched = (/**
         * @return {?}
         */
        () => {
        });
        this._onChange = (/**
         * @return {?}
         */
        () => {
        });
    }
    /**
     * @return {?}
     */
    get element() {
        return this._elementRef && this._elementRef.nativeElement;
    }
    /**
     * @return {?}
     */
    get format() {
        return this._format;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set format(value) {
        this._format = +value === 24 ? 24 : 12;
        /** @type {?} */
        const isDynamicallyChanged = value && (this._previousFormat && this._previousFormat !== this._format);
        if (isDynamicallyChanged) {
            this.value = this._value;
            this._timepicker.updateTime(this._value);
        }
        this._previousFormat = this._format;
    }
    /**
     * @return {?}
     */
    get max() {
        return this._max;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set max(value) {
        if (typeof value === "string") {
            this._max = NgxMatTimepickerAdapter.parseTime(value, { locale: this._locale, format: this.format });
            return;
        }
        this._max = value;
    }
    /**
     * @return {?}
     */
    get min() {
        return this._min;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set min(value) {
        if (typeof value === "string") {
            this._min = NgxMatTimepickerAdapter.parseTime(value, { locale: this._locale, format: this.format });
            return;
        }
        this._min = value;
    }
    /**
     * @param {?} picker
     * @return {?}
     */
    set timepicker(picker) {
        this._registerTimepicker(picker);
    }
    /**
     * @return {?}
     */
    get value() {
        if (!this._value) {
            return "";
        }
        return NgxMatTimepickerAdapter.toLocaleTimeString(this._value, { format: this.format, locale: this._locale });
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        if (!value) {
            this._value = "";
            this._updateInputValue();
            return;
        }
        /** @type {?} */
        const time = NgxMatTimepickerAdapter.formatTime(value, { locale: this._locale, format: this.format });
        /** @type {?} */
        const isAvailable = NgxMatTimepickerAdapter.isTimeAvailable(time, (/** @type {?} */ (this._min)), (/** @type {?} */ (this._max)), "minutes", this._timepicker.minutesGap, this._format);
        if (isAvailable) {
            this._value = time;
            this._updateInputValue();
            return;
        }
        console.warn("Selected time doesn't match min or max value");
    }
    /**
     * @private
     * @param {?} time
     * @return {?}
     */
    set _defaultTime(time) {
        this._timepicker.defaultTime = NgxMatTimepickerAdapter.formatTime(time, {
            locale: this._locale,
            format: this.format
        });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.value && changes.value.currentValue) {
            this._defaultTime = changes.value.currentValue;
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._unregisterTimepicker();
        this._subsCtrl$.next();
        this._subsCtrl$.complete();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        if (!this.disableClick) {
            this._timepicker.open();
            event.stopPropagation();
        }
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this._onChange = fn;
    }
    /**
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @param {?} isDisabled
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.disabled = isDisabled;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    updateValue(value) {
        this.value = value;
        this._onChange(value);
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.value = value;
        if (value) {
            this._defaultTime = value;
        }
    }
    /**
     * @private
     * @param {?} picker
     * @return {?}
     */
    _registerTimepicker(picker) {
        if (picker) {
            this._timepicker = picker;
            this._timepicker.registerInput(this);
            this._timepicker.timeSet
                .pipe(takeUntil(this._subsCtrl$))
                .subscribe((/**
             * @param {?} time
             * @return {?}
             */
            (time) => {
                this.value = time;
                this._onChange(this.value);
                this.onTouched();
                this._defaultTime = this._value;
            }));
        }
        else {
            throw new Error("NgxMatTimepickerComponent is not defined." +
                " Please make sure you passed the timepicker to ngxMatTimepicker directive");
        }
    }
    /**
     * @private
     * @return {?}
     */
    _unregisterTimepicker() {
        if (this._timepicker) {
            this._timepicker.unregisterInput();
        }
    }
    /**
     * @private
     * @return {?}
     */
    _updateInputValue() {
        this._elementRef.nativeElement.value = this.value;
    }
}
NgxMatTimepickerDirective.decorators = [
    { type: Directive, args: [{
                selector: "[ngxMatTimepicker]",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: NgxMatTimepickerDirective,
                        multi: true
                    }
                ],
                // tslint:disable-next-line:no-host-metadata-property
                host: {
                    "[disabled]": "disabled",
                    "(blur)": "onTouched()",
                },
            },] }
];
/** @nocollapse */
NgxMatTimepickerDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: MatFormField, decorators: [{ type: Optional }, { type: Inject, args: [MatFormField,] }] },
    { type: String, decorators: [{ type: Inject, args: [NGX_MAT_TIMEPICKER_LOCALE,] }] }
];
NgxMatTimepickerDirective.propDecorators = {
    format: [{ type: Input }],
    max: [{ type: Input }],
    min: [{ type: Input }],
    timepicker: [{ type: Input, args: ["ngxMatTimepicker",] }],
    value: [{ type: Input }],
    cdkOverlayOrigin: [{ type: HostBinding, args: ["attr.cdkOverlayOrigin",] }],
    disableClick: [{ type: Input }],
    disabled: [{ type: Input }],
    onClick: [{ type: HostListener, args: ["click", ["$event"],] }],
    updateValue: [{ type: HostListener, args: ["change", ["$event"],] }]
};
if (false) {
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.cdkOverlayOrigin;
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.disableClick;
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.disabled;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._format;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._max;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._min;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._previousFormat;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._subsCtrl$;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._timepicker;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._value;
    /** @type {?} */
    NgxMatTimepickerDirective.prototype.onTouched;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._onChange;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._elementRef;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._matFormField;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerDirective.prototype._locale;
}
