// ngx-mat-timepicker/esm2015/lib/directives/ngx-mat-timepicker-autofocus.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ngx-mat-timepicker-autofocus.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Inject, Input, Optional } from "@angular/core";
import { DOCUMENT } from "@angular/common";
export class NgxMatTimepickerAutofocusDirective {
    /**
     * @param {?} _element
     * @param {?} document
     */
    constructor(_element, document) {
        this._element = _element;
        this.document = document;
        this._activeElement = this.document.activeElement;
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        if (this.isFocusActive) {
            // To avoid ExpressionChangedAfterItHasBeenCheckedError;
            setTimeout((/**
             * @return {?}
             */
            () => this._element.nativeElement.focus({ preventScroll: true })));
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        // To avoid ExpressionChangedAfterItHasBeenCheckedError;
        setTimeout((/**
         * @return {?}
         */
        () => this._activeElement.focus({ preventScroll: true })));
    }
}
NgxMatTimepickerAutofocusDirective.decorators = [
    { type: Directive, args: [{
                selector: "[ngxMatTimepickerAutofocus]"
            },] }
];
/** @nocollapse */
NgxMatTimepickerAutofocusDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [DOCUMENT,] }] }
];
NgxMatTimepickerAutofocusDirective.propDecorators = {
    isFocusActive: [{ type: Input, args: ["ngxMatTimepickerAutofocus",] }]
};
if (false) {
    /** @type {?} */
    NgxMatTimepickerAutofocusDirective.prototype.isFocusActive;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerAutofocusDirective.prototype._activeElement;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerAutofocusDirective.prototype._element;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerAutofocusDirective.prototype.document;
}
