// ngx-mat-timepicker/esm5/lib/components/ngx-mat-timepicker-minutes-face/ngx-mat-timepicker-minutes-face.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ngx-mat-timepicker-minutes-face/ngx-mat-timepicker-minutes-face.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { NgxMatTimepickerUnits } from "../../models/ngx-mat-timepicker-units.enum";
import { NgxMatTimepickerPeriods } from "../../models/ngx-mat-timepicker-periods.enum";
import { NgxMatTimepickerUtils } from "../../utils/ngx-mat-timepicker.utils";
//
import { DateTime } from "ts-luxon";
var NgxMatTimepickerMinutesFaceComponent = /** @class */ (function () {
    function NgxMatTimepickerMinutesFaceComponent() {
        this.minuteChange = new EventEmitter();
        this.minutesList = [];
        this.timeUnit = NgxMatTimepickerUnits;
        this._color = "primary";
    }
    Object.defineProperty(NgxMatTimepickerMinutesFaceComponent.prototype, "color", {
        get: /**
         * @return {?}
         */
        function () {
            return this._color;
        },
        set: /**
         * @param {?} newValue
         * @return {?}
         */
        function (newValue) {
            this._color = newValue;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @param {?} changes
     * @return {?}
     */
    NgxMatTimepickerMinutesFaceComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.period && changes.period.currentValue) {
            /** @type {?} */
            var minutes = NgxMatTimepickerUtils.getMinutes(this.minutesGap);
            this.minutesList = NgxMatTimepickerUtils.disableMinutes(minutes, this.selectedHour, {
                min: this.minTime,
                max: this.maxTime,
                format: this.format,
                period: this.period
            });
        }
    };
    NgxMatTimepickerMinutesFaceComponent.decorators = [
        { type: Component, args: [{
                    selector: "ngx-mat-timepicker-minutes-face",
                    template: "<ngx-mat-timepicker-face [color]=\"color\"\n                     [faceTime]=\"minutesList\"\n                     [selectedTime]=\"selectedMinute\"\n                     [minutesGap]=\"minutesGap\"\n                     (timeChange)=\"minuteChange.next($event)\"\n                     [unit]=\"timeUnit.MINUTE\"></ngx-mat-timepicker-face>\n"
                }] }
    ];
    /** @nocollapse */
    NgxMatTimepickerMinutesFaceComponent.ctorParameters = function () { return []; };
    NgxMatTimepickerMinutesFaceComponent.propDecorators = {
        color: [{ type: Input }],
        format: [{ type: Input }],
        maxTime: [{ type: Input }],
        minTime: [{ type: Input }],
        minuteChange: [{ type: Output }],
        minutesGap: [{ type: Input }],
        period: [{ type: Input }],
        selectedHour: [{ type: Input }],
        selectedMinute: [{ type: Input }]
    };
    return NgxMatTimepickerMinutesFaceComponent;
}());
export { NgxMatTimepickerMinutesFaceComponent };
if (false) {
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.format;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.maxTime;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minTime;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minuteChange;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minutesGap;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.minutesList;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.period;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.selectedHour;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.selectedMinute;
    /** @type {?} */
    NgxMatTimepickerMinutesFaceComponent.prototype.timeUnit;
    /**
     * @type {?}
     * @private
     */
    NgxMatTimepickerMinutesFaceComponent.prototype._color;
}
